### ISSUE TITLE
SSH Tunneling is broken with SSH Agent with public key

### ISSUE THREAD

**Issue by mattpepin** [FreeRDP/Remmina#1228]
Version 1.2.0-rcgit-19 broke SSH tunneling using a public key with SSH Agent. My passphrase is never asked and no error message is displayed (no dialog box, no log message in a console window. Public key (automatic) works, but I don't want Remmina to manage my keys.

**Comment by antenore** [comment]
I'll take a look. I use solely SSH Agent, but I enter my password before the Remmina session, so I didn't realize there was this issue.

I won't have internet access and neither a keyboard for a couple of weeks, so maybe you can go back to rcgit-18.

**Comment by mattpepin** [comment]
No worries, I'll use an older package for now (I'm on Arch linux, but I found a question about this on Askubuntu: https://askubuntu.com/questions/939524/remmina-1-2-ssh-authentication-options).

In my case, it won't work even if I used my key earlier (passphrase in memory).

**Comment by kevin201** [comment]
I asked that question. I haven't been able to install rcgit-18, nor find any documentation on what the difference is between 'SSH Agent' and 'Public Key (Automatic)'.  I can get public key to work if I put the password into remmina, but SSH Agent just produces a "(null)" error message for me. Grateful for any pointers.

**Comment by box293** [comment]
I updated Linux Mint yesterday and I think this issue has hit me. I also get the Null message when trying to SSH.

However if I go and change from "SSH agent" to "SSH identity file" everything works ... although I would prefer not having to update all my existing connections.

**Comment by antenore** [comment]
We use libssh behind remmina and as you can see in the doc:

> Tries to automatically authenticate with public key and "none".  
> It may fail, for instance it doesn't ask for a password and uses a default asker for passphrases (in case the private key is encrypted). 

So it's the right behavior, in fact it works when the gnome keyring has stored the password, but it's quite confusing.

I'll investigate further if it's possible to simplify it or rollback it somehow.

To change all the profiles in one shot it's quite easy from the command line using sed (under  ~/.local/share/remmina), if you need a quick workaround and don't know how to do let me know.

**Comment by box293** [comment]
No worries.

This command is what I used:
```
sed -i 's/^ssh_auth=2/ssh_auth=1/g' *.remmina
```

The files were in `~/.remmina`

**Comment by kevin201** [comment]
>in fact it works when the gnome keyring has stored the password

I haven't been able to get this functionality so far. 

I can however ssh from the command line with no password interaction, having used 'ssh-add' (and entered the private key password) at logon.

It would be interesting to know the difference between 'SSH Agent' and the public key auto methods. If this is the desired behaviour maybe it's just a question of understanding the docs.

**Comment by antenore** [comment]
> The files were in ~/.remmina 
@box293 This is because you are an old remmina supporter ;-) (thanks)
 
@kevin201 it works for me, but I use keychain at logon time that automatically add my ssk keys.
With gnome_keyring should be the same, but must be configured to save ssh keys as well at logon time.
Sorry for the confusion.

It's still not clear for me as well how to use the auto method. I think that I've started to implement it in Remmina for the dual factor authentication, but I don't remember.

Still investiagting

**Comment by antenore** [comment]
I remember now and it should work correcthly.

ssh_userauth_publickey_auto() uses the SSH_AGENT and you pass to it the username and the passphrase (in the password field)
ssh_userauth_agent() uses the SSH_AGENT as well but you don't need to pass the passphrase (only the username).

So can you confirm that entering the passphrase in the password field with the auto method works?

I should find the time to update the user doc, but it's not really easy to find free time :-\ 

https://github.com/FreeRDP/Remmina/wiki/Remmina-User's-guide

**Comment by kevin201** [comment]
Thanks for this.

> ssh_userauth_publickey_auto() uses the SSH_AGENT and you pass to it the username and the passphrase (in the password field)

That certainly works. Auto and passphrase picks up the key and let's me connect.

> ssh_userauth_agent() uses the SSH_AGENT as well but you don't need to pass the passphrase (only the username).

I haven't been able to get this to work at all with my setup. (From the terminal, though, I can 'ssh my.other.machines.com' and be connected without further interaction. I add the key/passphrase at logon. I think this is using ssh_agent)

I am having trouble understanding the benefit in having both of those options though. Shouldn't the key just be accepted if it's in the key ring, and the user prompted for the passphrase if it's needed?

I will bring some more definitive debugging in case I can help more.

**Comment by antenore** [comment]
Wait! I've the same issue, my bad!!!
So I can debug as well.

Sorry for this, I've set all of my ssh profiles to use the auto method.

I'll see what happened.

**Comment by antenore** [comment]
Do you guys can test the last PR ( #1246 )  I've submitted?

**Comment by Liath** [comment]
Can confirm #1246 fixes this issue for me. Thanks @antenore. :)

**Comment by antenore** [comment]
Cool! We'll release a new release ASAP.

**Comment by kevin201** [comment]
I'd like  to test, but I'm afraid I don't have enough knowledge of git to accomplish.

**Comment by aperomsik** [comment]
I hope you'll post here when the PPA is updated. I switched back to the ubuntu packaged version due to this issue.

**Comment by Arkem5** [comment]
Well... we are in December now, and it is still impossible to use remmina because it ask me for a passphrase each time i try to connect with a key instead of a password, and I always forbid the password connection on my SSH servers. Is it planned to fix this someday?

**Comment by mattpepin** [comment]
Still not fixed in Arch Linux either.

**Comment by antenore** [comment]
Please help me to understand the problem, because I'm quite lost

I've just tested the last commit in our main branch (named 'next' https://github.com/FreeRDP/Remmina/commit/202d17f5bec8d8aac386b647bf5f4be3c2ab0ff7 ) and it works.

My config:

Protocol: SSH
Server: barce (my final destination) 192.168.122.167:22
Tunnel: rex (my stepping stone)  192.168.0.28:22
Authentication Type: SSH Agent

I've an SSH Agent running through keychain.

The Public Key auto method means:

> The function ssh_userauth_autopubkey() does this using the available keys in "~/.ssh/". The return values are the following:
> 
>     SSH_AUTH_ERROR: some serious error happened during authentication
>     SSH_AUTH_DENIED: no key matched
>     SSH_AUTH_SUCCESS: you are now authenticated
>     SSH_AUTH_PARTIAL: some key matched but you still have to provide an other mean of authentication (like a password).
> 
> The ssh_userauth_publickey_auto() function also tries to authenticate using the SSH agent, if you have one running, or the "none" method otherwise. 

Ref: [ssh_userauth_publickey_auto](http://api.libssh.org/stable/group__libssh__auth.html#ga53e6771b250c061463ed98b6e5b6e0af)

For the pure ssh_agent implementation, the one I've just tested successfully, as far as I understand from the documentation, needs a local ssh agent running, it's not Remmina (read libssh in this case) that will execute one, nor Remmina will catch if the agent needs the authentication.

[ssh_userauth_agent](http://api.libssh.org/stable/group__libssh__auth.html#ga7d84f3f89f65455b80b10e2643d80719) 

What am I missing? Looking at the old code I don't see anything that could start the agent automatically.

If the 'next' branch works, on the other hand, I still need some time to publish a new release, because I'm working on a couple of things that have to be done.

As a temporary workaround, if you have one of the last commits, you should have an interface like this:

![screenshot_2017-12-15_23-51-54](https://user-images.githubusercontent.com/5975/34063571-2770bd96-e1f3-11e7-9be8-7f15412c6fd7.png)

It's to set an ssh tunnel in the prior command, i.e. I've added this in a script:

ssh -N -L 2222:192.168.122.167:22 -i /home/myusername/.ssh/id_rsa thatusername@192.168.100.28 &

Plus other commands to log the IP of the tunnel.

Than in the "post command" I've an other script path that kill the ssh_agent at the end of the connection.

The prior and post commands now supports niddles (%u %U %h %t) and command arguments, so I think you have several means to temporary workaround this issue. 

But please, help me to identify the exact problem if it's still there and sorry...

**Comment by antenore** [comment]
TL;DR
To resume, I've just saw a couple of comments.

This issue is solved, but unlucky, I cannot release a new version right now.

On Arch you can install the remmina-git AUR package, the package name include the version number, but it's worng, the AUR is always built from the last commit, so go ahead with it.

Ubuntu has the remmina-next-daily , follow this https://github.com/FreeRDP/Remmina/wiki/Testing-under-Ubuntu-with-the-remmina-next-daily-PPA

If with the 2 repositories above you still have issues, read my previous comment and help me :-)

Thanks

**Comment by mattpepin** [comment]
Hmmm nope... I just tried the latest GIT, same issue unfortunately!

**Comment by mattpepin** [comment]
Here's my profile, with sensitive info edited. I get the error message I mentioned above. It still does not work if I SSH from the command line before to unlock my key. It works if I use Public key (automatic), but I don't want Remmina to manage my keys.

`[remmina]
sound=off
sharefolder=
microphone=0
name=Test RDP
proxy=
cert_ignore=0
ssh_enabled=1
showcursor=0
exec=
clientname=
console=0
colordepth=16
server=test-pc
keyboard_grab=1
loadbalanceinfo=
ssh_auth=2
scale=1
aspectscale=0
ssh_charset=
quality=2
disableencryption=0
ssh_loopback=0
group=TESTING
ssh_username=server
password=test
username=test
hscale=0
disablepasswordstoring=0
gateway_server=
window_maximize=1
viewmode=1
gateway_usage=0
disableautoreconnect=0
viewonly=0
postcommand=
resolution=
security=
window_height=704
keymap=
domain=
precommand=
window_width=1364
sharesmartcard=0
ssh_server=testing.duckdns.org:2642
protocol=RDP
shareprinter=0
disableserverinput=0
disableclipboard=0
ssh_privatekey=
vscale=0
execpath=`

**Comment by mattpepin** [comment]
My profiles are in ~/.remmina, not in ~/.local/share/remmina. I guess it's distro specific? I used the path shown in the status bar.

**Comment by antenore** [comment]
> My profiles are in ~/.remmina, not in ~/.local/share/remmina. I guess it's distro specific? I used the path shown in the status bar. 

It's not distro specific, ~/.remmina comes from an old version of remmina. The new versions of Remmina, new as about year old maximum, copies automatically the profiles from ~/.remmina to ~/.local/share/remmina.
Can you check if you still have an old version installed somehow?
The problem may lie in a melange of remmina versions.

**Comment by Arkem5** [comment]
Maybe I have a clue...
I can connect to a ssh server on a Mint 18.3 ssh server using RSA key at work.
In the other cases, it doesn't work either with Debian Jessie and Mint 18.3 ssh servers using ed25519 keys. Maybe there is an issue with this one, but I expected the agent to handle it all...
Connections using the ssh shell or vinagre works normally in all cases. I prefer Remmina for VNC with SSH encryption, though, because Vinagre's interface is missing too many important things...

**Comment by antenore** [comment]
@Arkem5 this is cool! Thank you! Something I can work out

**Comment by mattpepin** [comment]
You were right, my profiles were also copied to .local/share/remmina, but for some reason, it still used the old ones. I tried removing everything from both locations and start from scratch, but the issue persists. I can also confirm I only have 1 remmina binary (which is in /usr/bin).

**Comment by antenore** [comment]
Please TEST. Hopefully I've just pushed a fix for this.
As per libssh documentation we SHOULD not send the username to the ssh_agent.
In some cases, I think, when the local and remote users are the same, or when in the ~/.ssh/config file the user name is set, the ssh_agent was working (I was in these 2 cases).

I'm quite sure I've broken something else, so please test.

Thank you for our patience!!!

**Comment by Arkem5** [comment]
I just tried the Ubuntu next-daily version, but I still get this message: kex error : no match for method server host key algo: server [ssh-ed25519], client [ecdsa-sha2-nistp256,ssh-rsa,ssh-dss]
Do I have to compile the git to have a more up to date version?

**Comment by mattpepin** [comment]
I just recompiled latest git (git 055e33b33), and I still have the issue.

**Comment by antenore** [comment]
I'm not sure that the remmina daily already contains my fixes, it'll for sure tomorrow.
On the other side, I think that you have a different issue and I may have a solution.
Would you mind opening a new issue for that?
Thanks!

**Comment by antenore** [comment]
Yep, as I was saying, you have a different problem, let's solve it outside of this thread.
Can you test an ssh connection specifying the cypher manually in Remmina?

**Comment by mattpepin** [comment]
In my case, I know libssh is parsing my config file (~/.ssh/config), because if I create an alias and use it in Remmina, Remmina connects to the right server. I'm not sure what's causing the error message.

**Comment by Arkem5** [comment]
You are right, this one was different. I didn't notice the error message isn't the same when connecting to a debian ssh server. I will make another issue for it if it persists...
I just tried again with a mint server and I get the error "Access denied. Authentication that can continue: publickey" again. I guess I should compile instead of using the next-daily, tough...

**Comment by antenore** [comment]
I'm mixing up the problems @mattpepin , sorry.

**Comment by mattpepin** [comment]
No problem, I'm just trying to figure out what's not working. I made some experiments to check if the SSH config file caused issues, but it seems not. I don't know why you can't replicate.

**Comment by antenore** [comment]
One thing got me crazy. The private keys must have the same name root, like foobar and foobar.pub

This is very important

**Comment by mattpepin** [comment]
Yes, I'm using the default id_rsa and id_rsa.pub in my ~/.ssh folder.

**Comment by antenore** [comment]
I'm short of ideas.
https://blog.codefront.net/2007/02/28/debugging-ssh-public-key-authentication-problems/

On Remmina you set the ssh verbosity in the general preferences.

The sshd test may really help out and the authlog as well.

**Comment by antenore** [comment]
Ah, I forgot to mention to open the Remmina debug window when you test with a different verbosity.

**Comment by mattpepin** [comment]
Here's the end of the debug log:

[SSH] ssh_agent_get_ident_count: Answer type: 12, expected answer: 12
[SSH] ssh_agent_get_ident_count: Agent count: 2
[SSH] pki_import_pubkey_buffer: Invalid public key length
[SSH] ssh_socket_unbuffered_write: Enabling POLLOUT for socket
[SSH] packet_send2: packet: wrote [len=28,padding=11,comp=16,payload=16]

**Comment by mattpepin** [comment]
My private key is 4096 bytes long, not 2048. Maybe that's the issue?

**Comment by Arkem5** [comment]
Here's my debug log with the standard Remmina version (I didn't have time to compile for now)

[SSH] ssh_config_parse_file: Reading configuration data from /etc/ssh/ssh_config
[SSH] ssh_config_parse_line: Unsupported option: SendEnv, line: 53
[SSH] ssh_config_parse_line: Unsupported option: HashKnownHosts, line: 54
[SSH] ssh_config_parse_line: Unsupported option: GSSAPIAuthentication, line: 55
[SSH] ssh_connect: libssh 0.6.3 (c) 2003-2014 Aris Adamantiadis, Andreas Schneider, and libssh contributors. Distributed under the LGPL, please refer to COPYING file for information about your rights, using threading threads_pthread
[SSH] socket_callback_connected: Socket connection callback: 1 (0)
[SSH] ssh_client_connection_callback: SSH server banner: SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.2
[SSH] ssh_analyze_banner: Analyzing banner: SSH-2.0-OpenSSH_7.2p2 Ubuntu-4ubuntu2.2
[SSH] ssh_analyze_banner: We are talking to an OpenSSH client version: 7.2 (70200)
[SSH] ssh_packet_userauth_failure: Access denied. Authentication that can continue: publickey
[SSH] ssh_agent_get_ident_count: Answer type: 12, expected answer: 12
[SSH] ssh_packet_userauth_failure: Access denied. Authentication that can continue: publickey
[SSH] ssh_packet_userauth_failure: Access denied. Authentication that can continue: publickey

Here are the 53 - 55 lines from my ssh_config:
    SendEnv LANG LC_*
    HashKnownHosts yes
    GSSAPIAuthentication yes

**Comment by antenore** [comment]
> My private key is 4096 bytes long, not 2048. Maybe that's the issue? 

@mattpepin Looks like to me that you have one of the following issue: 

1. An ssh agent running with an old (2048) key loaded
Check you don't have any other ssh-agent (keychain, gnome-keyring, gpg-agent, ???)
2. Your 2048 key is sent as the first key and when it fails libssh doesn't try the second one
libssh, as far as I've understood, should send one key after an other, till it gets a positive answer or till there are not any new keys. If this is the case, I think it's more a libssh issue.
Try to unload this key and try again, as you have an older version of libssh, I don't expect they will fix this behaviour.
3. A libssh bug
It could be, but I don't know how we can debug this, except writing a libssh ssh client for debugging purpose.

I hope in the 1st case honesty, I'm honestly puzzled

**Comment by antenore** [comment]
> I just tried the Ubuntu next-daily version, but I still get this message: kex error : no match for method server host key algo: server [ssh-ed25519], client [ecdsa-sha2-nistp256,ssh-rsa,ssh-dss]
> Do I have to compile the git to have a more up to date version? 

@Arkem5 Is this error related to the SSH debug messages you have shared?

In this case you should be able to workaround it setting the Kex in the ~/.ssh/config file

```
KexAlgorithms ssh-ed25519
``` 
Let me know if it helps.

If you have the last remmina-daily, I've added some new feature in the SSH protocol (not in the agent, but in the protocol itself). Would you mind trying it specifying the correct algorithms? If it works I could add that options in the ssh tunnel tab

**Comment by mattpepin** [comment]
It turns out it does have a link with my private keys. If I only leave my id_rsa and id_rsa.pub in my ~/.ssh folder, it works! I iterated over the keys I had, and I found out that the culprit was id_ecdsa and id_ecdsa.pub. As soon as I remove this key, it works. I did not even use it anymore, as I moved to ed25519. The issue is resolved, but I still think it's a bug because the login should have still worked.

**Comment by antenore** [comment]
@mattpepin this is because of #1187 I don't know how I can make everybody happy :-P 
I'll take a look if I can improve this

**Comment by Arkem5** [comment]
Here's what I did on my side 
1. I removed everything related to Remmina, then I installed from Git, and got the 1.2.0-rcgit-24 (git 055e33b) version
2. I created a ~/.ssh/config file, and copied the line you sent me inside (I never needed such file for connecting through SSH until now
3. I added the string ed25519 in the Advanced -> KEX (Key Exchange) Algorythm entry for both Mint and Debian SSH servers (wich use the same ed25519 key). Note that for now, it is only SSH access through Remmina, not tunneling. I hope it is fine (SSH connection through SSH Tunnel is also deactivated in the third tab)
4. I ensured my client's .ssh folder and all its content's permissions give only access to user in case of...
5. I removed all other dsa, ecdsa and rsa keys I don't use anymore from the folder since it seemed to be useful for Mattpepin

Sadly, the results are still the same, and I still get 2 different errors depending on the server's distribution:
* Mint 18.3 (amd_64) : Access denied. Authentication that can continue: publickey
* Debian Jessy (arm) : kex error : no match for method server host key algo: server [ssh-ed25519], client [ecdsa-sha2-nistp256,ssh-rsa,ssh-dss]

Logging with the shell didn't work anymore with the ~/.ssh/config file. I got the error :

> Unsupported KEX algorithm "ssh-ed25519"
/home/arkem/.ssh/config line 1: Bad SSH2 KexAlgorithms 'ssh-ed25519'.

I tried changing it to 'ed25519' only, but nothing changed, so I removed the file and I could log again to both servers using the shell, but nothing changes with Remmina...

**Comment by antenore** [comment]
@Arkem5 sorry, my bad, I've given you the wrong Kex, let me search and I'll give you the right one.

The one supported by your client and your server can be listed with ```ssh -Q kex```

**Comment by antenore** [comment]
@Arkem5 

I've tested an ed25512 key and it works

1. I've created  and installed the key

```sh
$ ssh-keygen -t ed25519 -f ~/.ssh/id_ed25519 -C "Remmina - test ed25519"
$ ssh-copy-id -i .ssh/id_ed25519 tmow@rex
```

2. Tested by hand (works)

3. Created the Remmina profile
(Here after the relevant options)

```
[remmina]
ssh_auth=2
ssh_privatekey=/home/tmow/.ssh/id_ed25519
ssh_color_scheme=0
ssh_loopback=0
name=rex ed25519
ssh_username=tmow
protocol=SSH
sshlogenabled=0
ssh_enabled=0
ssh_stricthostkeycheck=0
disablepasswordstoring=0
server=rex
```
5. Removed unneeded ssh keys:

```sh
ssh-add -l
ssh-add -d /path/to/some_key
```
6. Test Remmina

Works!

I'm on Arch, so I've only the last version of almost everything.

**Comment by antenore** [comment]
@Arkem5 also tested an RDP connection through an SSH tunnel, and it works.

Do you have a distro with an updated libssh version?

**Comment by antenore** [comment]
> It turns out it does have a link with my private keys. If I only leave my id_rsa and id_rsa.pub in my ~/.ssh folder, it works! I iterated over the keys I had, and I found out that the culprit was id_ecdsa and id_ecdsa.pub. As soon as I remove this key, it works. I did not even use it anymore, as I moved to ed25519. The issue is resolved, but I still think it's a bug because the login should have still worked.

@mattpepin I'm going to push a commit that should fix this bad behaviour, tell me what do you think about the logic

1. Remmina first searches for identities in the ~/.ssh folder in this order

```
.ssh/id_ed25519
.ssh/id_rsa
.ssh/id_dsa
.ssh/identity
```
The first identity file (private key) found id the only identity that will be used.
2. Identity file explicitly set in the profile settings.
If the user has set in the profile option an identity file, that identity file will be used for any kind of public key authentication (ssh-agent, automatic, etc)

In this way all the cases should be covered.

@ssten I'd love to have also your opinion on this whole thread, do you mind? Thanks

**Comment by Arkem5** [comment]
I tried the command ssh -Q kex on all (client and servers) machines
- Mint 18.3 client and server replies:
diffie-hellman-group1-sha1
diffie-hellman-group14-sha1
diffie-hellman-group-exchange-sha1
diffie-hellman-group-exchange-sha256
ecdh-sha2-nistp256
ecdh-sha2-nistp384
ecdh-sha2-nistp521
curve25519-sha256@libssh.org
- Debian Jessie replies the same plus:
diffie-hellman-group1-sha1

I tried giving the path to my client key (/home/marc/.ssh/id_ed25519) to the Remmina configuration, but nothing changed

Note that in my case, the user name is not the same on both client and server machines and I only allow one specific user to be accessed through ssh on the server. also, I never use the standard port 22, but something different on each server... could you try with different client and server user names on your side?

**Comment by antenore** [comment]
It's what I do everyday, but I've the users set in the ~/ssh/config file 
I'll try without parsing the config file.

**Comment by ssten** [comment]
Accordingly with (#1374 ) **~/.ssh/id_dsa** *may* be used, but in such case user **must** be warned about usage of obsolete algorythm.
**~/.ssh/identity** is the thing I've just *heard*, but don't used and incompletely understand.

Commonly, for my experience, attepmts to guess proper private key may provide uclear to understand and unpleasant surprises.
Documented list of checked private key files is a good and right thing. But it is not complete.
I suggest to check number of found standard private keys. And, if find more than one, autoselect mode should fail, notifying user, that he **must** select exact one.

**Comment by antenore** [comment]
I'm pushing these changes in next.

ASAP we will do a release.

**Comment by mattpepin** [comment]
Just rebuilt my package. Seems fine! One minor thing though, when I connect, the agent asks for my passphrase (normal). If I cancel the prompt, the error message is "out of memory". It's a minor issue and it might have been this way before too...

**Comment by antenore** [comment]
I've released rcgit.25 in the meanwhile
